### PR TITLE
Split legalization pass to consider IR and features separately.

### LINKED ISSUE #7158: objmode error in IR legalization
## Reporting a bug

- [x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

Looks like IR legalization is called on objmode IR, potentially throwing error for valid code. Is IR legalization pass needed for objmode IR in general?
```python
import numba
import numpy as np

@numba.njit
def f(A):
    with numba.objmode(B="int64[::1]"):
        B = A.view("int64")
    return B

f(np.ones(11, np.int64))
```
```
  File "/Users/ehsan/dev/numba/numba/core/typeinfer.py", line 1071, in propagate
    raise errors[0]
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: Handle with contexts)
Failed in nopython mode pipeline (step: nopython frontend)
Failed in object mode pipeline (step: ensure IR is legal prior to lowering)
'view' can only be called on NumPy dtypes, try wrapping the variable 'A' with 'np.<dtype>()'

File "../tmp/objmode_ir_legal.py", line 7:
def f(A):
    <source elided>
    with numba.objmode(B="int64[::1]"):
        B = A.view("int64")
        ^

During: resolving callee type: type(ObjModeLiftedWith(<function f at 0x10180d040>))
During: typing of call at /Users/ehsan/dev/bodo/../tmp/objmode_ir_legal.py (6)


File "../tmp/objmode_ir_legal.py", line 6:
def f(A):
    with numba.objmode(B="int64[::1]"):
    ^
```

### PR DESCRIPTION
**Pull request by stuartarchibald**
This splits out legalization into:
* IR legalization
* Feature legalization for nopython mode and object mode.

It puts these into separate compiler passes and adds them to the
standard pipelines.

Adds test based on original reproducer.

Fixes #7158

### CODE REVIEW

File: numba/core/typed_passes.py [hunk 1]
```diff
@@ -437,16 +437,45 @@ def run_pass(self, state):
 
 
 @register_pass(mutates_CFG=False, analysis_only=True)
-class IRLegalization(AnalysisPass):
+class NoPythonSupportedFeatureValidation(AnalysisPass):
+    """NoPython Mode check: Validates the IR to ensure that features in use are
+    in a form that is supported"""
 
-    _name = "ir_legalization"
+    _name = "no_python_supported_feature_validation"
```
Comment (sklam) [inline numba/core/typed_passes.py:444]:
```suggestion
    _name = "nopython_supported_feature_validation"
```

we always refer to it as one word; else it reads like it not doing validation

File: numba/core/typed_passes.py [hunk 2]
```diff
@@ -437,16 +437,45 @@ def run_pass(self, state):
...
         warn_deprecated(state.func_ir, state.typemap)
+        return False
+
+
+@register_pass(mutates_CFG=False, analysis_only=True)
+class ObjModeSupportedFeatureValidation(AnalysisPass):
+    """Object mode check: Validates the IR to ensure that features in use are
+    in a form that is supported"""
+
+    _name = "obj_mode_supported_feature_validation"
```
Comment (sklam) [inline numba/core/typed_passes.py:460]:
```suggestion
    _name = "objmode_supported_feature_validation"
```
one word is better?

File: numba/core/typed_passes.py [hunk 3]
```diff
@@ -437,16 +437,45 @@ def run_pass(self, state):
...
+
+    def __init__(self):
+        AnalysisPass.__init__(self)
+
+    def run_pass(self, state):
+        warn_deprecated(state.func_ir, state.typemap)
```
Comment (sklam) [inline numba/core/typed_passes.py:466]: I don't think `warn_deprecated` is needed either. All the deprecated feature are nopython features, and `typemap` is all `pyobject`.

**Review by sklam — COMMENTED** [review]
Thanks for the patch. I have a few comments.

File: numba/core/typed_passes.py (hunk 3 above)
Comment (stuartarchibald) [reply numba/core/typed_passes.py:466]: I think this is also likely true. Obj mode blocks can't return reflected containers and obj mode itself doesn't need to worry about it. Given this pass is now a NOP, remove it?

Comment (sklam) [reply numba/core/typed_passes.py:466]: yes remove it

File: numba/core/typed_passes.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/core/typed_passes.py]: indeed, fixed!

File: numba/core/typed_passes.py (hunk 3 above)
Comment (stuartarchibald) [reply numba/core/typed_passes.py]: done in 8c8222a

**Review by sklam — APPROVED** [review]
Thanks for the patch!

**Pull request merged**